transformer_engine/pytorch/module.py

```diff
@@ -135,6 +135,44 @@ def _prepare_backward(fp8: bool,
         delete_key_from_amax_buffer(forward=False)
 
 
+class NoopCat(torch.autograd.Function):
+    """This class is a no-op replacement for `torch.cat`."""
+
+    @staticmethod
+    def forward(ctx, full_param_buffer, *params_split):
+        assert not full_param_buffer.requires_grad, "Buffers should not require gradient"
+        assert (
+            full_param_buffer.shape[0] % len(params_split) == 0
+        ), "Dimensions not compatible for concatenation"
+
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        for i, p in enumerate(params_split):
+            p.data = full_param_buffer[i * split_size : (i+1) * split_size]
+
+        param_temp = full_param_buffer.new()
+        param_temp.set_(full_param_buffer.storage(),
+                        full_param_buffer.storage_offset(),
+                        full_param_buffer.size(),
+                        full_param_buffer.stride())
+        param_temp.requires_grad = True
+
+        ctx.save_for_backward(full_param_buffer, *params_split)
+        return param_temp
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        full_param_buffer, *params_split = ctx.saved_tensors
+
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        grads = []
+
+        for i, p in enumerate(params_split):
+            p.data = full_param_buffer[i * split_size : (i+1) * split_size]
```

Comment:
This is the same as the logic in forward? Why?

Reply:
You're right in the sense that for our use case it doesn't matter since we initialize the individual params from the full buffer's memory. Having this just keeps the `NoopCat` function more general and doesn't assume the params are already correctly populated, but if we don't care about that then we can remove it.

Reply:
Hm? But it is not what that does - it just sets the pointers, not actually copying the data?

Reply:
For correct functionality of fwd/bwd/checkpointing that's all we need right, explicit copies aren't required

Reply:
But my point is that those pointers should already be correct, otherwise the result is wrong, right? Or am I missing something?

Reply:
In the general use case for `NoopCat`. If `params_split` occupy different memory region from `full_param_buffer`, then we would need to set them explicitly, right?